### PR TITLE
Implement std threads support

### PR DESCRIPTION
**Pull request by AzureMarker**
Requires:
- [x] https://github.com/Meziu/libc/pull/5
- [x] https://github.com/Meziu/pthread-3ds/pull/9
- [x] https://github.com/Meziu/rust-linker-fix-3ds/pull/11
- [x] https://github.com/Meziu/libc/pull/7
- [x] https://github.com/Meziu/pthread-3ds/pull/10

### CODE REVIEW

File: library/std/src/thread/mod.rs
Comment (Meziu) [inline library/std/src/thread/mod.rs]: Do not call libctru functions by name. Remove direct references to libctru specific functionality.

**Review by Meziu — CHANGES_REQUESTED** [review]

**Comment by Meziu** [comment]
Also, while the function docs are fine, add an explanation of _why_ we need those implemented.

File: library/std/src/thread/mod.rs
Comment (Meziu) [inline library/std/src/thread/mod.rs]: This and the documentation about the core#3 aren’t necessary. This explanation is outside Rust’s std scope and only really matter when programming the CFW. Just for the sake of simplicity.

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [inline library/std/src/sys/unix/thread.rs]: Instead of adding args like this, would it make sense to use something like [an extension trait](
https://doc.rust-lang.org/std/os/unix/thread/trait.JoinHandleExt.html)? 

I worry that this kind of modification to the APIs may not be accepted upstream

File: library/std/src/thread/mod.rs
Comment (ian-h-chamberlain) [inline library/std/src/thread/mod.rs]: Should this be `unstable(feature = "")` ??

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [inline library/std/src/sys/unix/thread.rs]: Instead of asserts maybe just return `Err` for failure here? I think there's a helper function to convert `libc` calls into `io::Result` somewhere in std...

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [inline library/std/src/sys/unix/thread.rs]: Any specific reason this was chosen vs the other defaults already listed here? It seems a bit small, but I can't really tell how easy it is to bump against the limit... For that matter I wonder why they don't generally use `PTHREAD_STACK_MIN` (because rust threads need more memory than pthread by default does?).

**Review by ian-h-chamberlain — APPROVED** [review]

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [reply library/std/src/sys/unix/thread.rs]: Found [this](https://github.com/Meziu/rust-horizon/blob/29f86efb69ea518a4bd992eb586e7134250c6401/library/std/src/sys/unix/thread.rs#L675), which I guess means they default to this anyway for unknown targets?

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [inline library/std/src/sys/unix/thread.rs]: Would it make sense to use this [existing API](https://man7.org/linux/man-pages/man3/pthread_attr_setschedparam.3.html) which is defined on some platforms?

Edit: [and this one](https://man7.org/linux/man-pages/man3/pthread_attr_setaffinity_np.3.html) for affinity? That one is specifically non-portable I guess

File: library/std/src/sys/unix/thread.rs
Comment (Meziu) [reply library/std/src/sys/unix/thread.rs]: This looks like a better idea. Changing the api directly isn’t ideal, but this way we could have our own little module to do stuff like it.

File: library/std/src/thread/mod.rs
Comment (AzureMarker) [reply library/std/src/thread/mod.rs]: Sure, I just copied these docs from the ctru-rs module, but sounds like they need some changes.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: I thought about this, and that's how I expected it to work, but the issue is we need to now store more data in the builder. Since we're already directly modifying the builder, it made sense to keep the changes together here.

I could try a few things though to achieve an extension trait.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: This is how most other libc calls are being handled, but I could see about returning the error.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: This was the default used in ctru-rs, and it seems to show up a few times in libctru (but not directly for normal threads). We could change it if we want.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: I can try to route things through those interfaces.

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [reply library/std/src/sys/unix/thread.rs]: Ok, if it matches other calls to `libc::pthread_*` then consistency is probably best. I don't think we expect these calls to typically fail anyway...

File: library/std/src/sys/unix/thread.rs
Comment (ian-h-chamberlain) [reply library/std/src/sys/unix/thread.rs]: The main reason I brought this up is, I know from the tokio example you posted and my brief try using threads for the HashMap tests, the default `ctru` size was not large enough to do much useful, so it seems like we'd probably want to pick a usable default so most users wouldn't need to think about it. Whatever you think makes sense is fine.

Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: Since the std thread docs mention 2MiB is the default thread stack size, I can use that default here in std.

**Comment by AzureMarker** [comment]
> Also, while the function docs are fine, add an explanation of _why_ we need those implemented.

I added a little bit, but can you give an example of what you're thinking?

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: I did some refactoring and got an extension trait to work.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: It would be a programming bug if the affinity is wrong, and probably same for priority, so I'll leave it as an assert (esp. since most of the other pthread calls do this, probably for the same reasons).

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: If this fails we could also make our pthread extensions more noticably non-portable by adding the `_np` suffix like those affinity functions do.

Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: There's some interesting functions here: https://man7.org/linux/man-pages/man7/sched.7.html
But maybe these would be more for getting the priority and affinity values.

File: library/std/src/sys/sgx/thread.rs
Comment (Meziu) [inline library/std/src/sys/sgx/thread.rs:110]: Why are you modifying all of these system implementations?

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: It's to support the trait extension. We need to define extra fields on the builder struct, and we want to avoid adding 3ds specific cfg sections to that code. So we need to give all operating systems the chance to add new fields to the builder struct. And how they use the new fields is also platform specific, so that needs to happen here. Thus, the data is passed even if it's not used (it would be optimized out).

Comment (Meziu) [reply library/std/src/sys/sgx/thread.rs:110]: This is a big problem though. We should keep ourselves on `unix` only, and that’s already much. Changing all of these implementations isn’t something we can just push.
This is basically a whole new feature, it’s barely related to our “new target” changes at all.

Comment (Meziu) [reply library/std/src/sys/sgx/thread.rs:110]: We should take this to a _completely pthread_ level. We could use a global Mutex which holds the data of the “latest” building thread. It sounds absolutely trash, but we wouldn’t have to worry about `std` changes breaking the system. (We should flush the values stored in the mutex at thread creation, and keep everything in `pthread-3ds`, complete of renaming the functions to the standard declarations).

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: That is absolutely not going to work. You can have builders for two different threads active at the same time and interleave setting their options, which would break your idea. Even worse, it would utterly confuse the user since it breaks the API contract.

If we want an extension trait, and we want to do it right, we need to do this. I don't see what the big issue is. It doesn't change anything about how the other systems work. Since this is a new thing to std (being able to set target specific options on the thread builder), it shouldn't be unexpected that this kind of refactor would be necessary.

If std folks have a better idea about how to do this when we open the PR, then we can change it then. But we should not use hacks like that.

Comment (Meziu) [reply library/std/src/sys/sgx/thread.rs:110]: Thing is, we can’t just make random changes and expect the “std folks” to tell us what to do. These changes aren’t about our target anymore, this is modifying the public interface. We should first of all rename the libc functions to have the standard declarations, and then search for a new way to do this, one without changing the basics of the std, not all in one PR at least. Is there really no other way?

The whole point of our stubs, implementations, compatibility layers and such, is to provide a stable and durable interface to our target’s system. This change goes beyond that principle.

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: 1. These aren't random changes. We are doing a new thing here by allowing target specific thread builder properties. We don't want to hard code this for the 3DS only, so we need to give all platforms the ability to do this too. The other platforms don't use it right now though, so nothing changes functionally except for the 3DS.
2. This is in _absolutely no way_ modifying the public interface of the other platforms. There is no public change. These changes are all in private implementation modules. Maybe this is what you are tripping up on. There is no functional change to the other platforms. The `sys` module is not public.

Comment (Meziu) [reply library/std/src/sys/sgx/thread.rs:110]: I worded it wrongly, by "public interface", I meant the one `sys` modules provide to the `thread` module. I meant changing how each `sys` module is supposed to work, just to accomodate for a feature only we need.

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: Is it a feature we only use, but you guys asked for a trait extension (which is probably the right approach). You can't write a trait extension that adds fields to a struct, so we need a refactoring like this which supports adding fields to the thread builder in a target specific way.

At that point, we still don't _need_ to modify the other platforms, but we'd be hard coding it so only the 3DS is allowed to have custom thread builder properties. Based on what I saw in the other platforms' code, they would benefit from having this functionality too. Additionally, the point of having an imp module is to treat all platforms the same at a high level. That's why there's stuff like the imp::guard module defined in all platforms' code even though only a few support it. To keep the interface between the high level thread module and the platform specific modules stable/consistent, we need to implement no-op builder option support in all other platforms.

There's a review process for std PRs for a reason. If they have an alternative way of doing this and would rather us do it differently, they will tell us.

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: Don't be afraid of changing the sys module slightly. The whole point of that module is to be an internal (not stable) way to abstract over the various supported platforms. We can change it without fear.

Comment (ian-h-chamberlain) [reply library/std/src/sys/sgx/thread.rs:110]: I think @AzureMarker makes some good points here, and I'm not entirely opposed to a change set like this, but I do still have a few concerns. I haven't really contributed to `std` before, so I don't know too much about the process.

* Should we try to upstream smaller, less invasive changes (like most of the `#[cfg(...)]` etc.) before trying to add new features like this? It seems to me that maintainers might prefer to review things in smaller digestible chunks.
* Is this the kind of thing that an RFC (or [a MCP](https://forge.rust-lang.org/compiler/mcp.html)) would be required for? It's a fairly small feature, but it's still a feature that lives behind a feature gate. If so, I definitely think it would be preferable to separate it from the rest of our std changes since it would require a tracking issue, etc.

I'm slightly wary of building up too many big code changes in this repo and trying to get them all merged at once, and separating them out later will probably be painful. Other than trying to upstream sooner, I'm not sure we have a very good way to gauge how maintainers feel about stuff like this.

Edit: Found [some guidelines](https://github.com/rust-lang/rfcs/blob/master/libs_changes.md) for when things in `std` or libs need an RFC.

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: For the RFC question, I think we're fine to just do a PR since the public change (adding the trait extension) is behind an unstable feature flag. This is called out as a path in the guidelines you linked. We would still need an FCP or RFC to stabilize the feature.

Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: We could also make a Zulip thread (or wherever std people hang out) to get their opinion on it.

Comment (ian-h-chamberlain) [reply library/std/src/sys/sgx/thread.rs:110]: I've written up a question on Zulip: https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/std.3A.3Athread.20support.20for.20armv6k-nintendo-3ds/near/271654826

Feel free to add any additional context you think might be helpful, and we'll see if we get any feedback.

**Comment by AzureMarker** [comment]
@ian-h-chamberlain based on your comment here I did some poking around and it _may_ be possible to use libctru APIs to set the priority and affinity after-the-fact:
https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/std.3A.3Athread.20support.20for.20armv6k-nintendo-3ds

I'll look into supporting that in std, but we still may want the current API + the ability to set it after thread spawn.

**Comment by AzureMarker** [comment]
These are some interesting wiki pages:
https://www.3dbrew.org/wiki/Multi-threading
https://www.3dbrew.org/wiki/KThread

From this, we should change how we refer to the core ID we pass during thread creation. "Affinity" usually refers to a CPU affinity bit mask that tells the kernel which cores a thread could run on (or a list of masks, for multiple CPUs). The 3DS wiki refers to the core ID passed during thread creation as the "ideal processor" ID. The 3DS kernel stores both pieces of information (see offsets 0x7E and 0x90 on the KThread page).

The affinity mask concept is pretty common, so there are some existing pthread APIs we could reuse. Similar for thread priority by the way. But the ideal processor idea is not so common, so I think we will still need our own nonstandard pthread function for that. Based on the naming scheme used below, I nominate `pthread_attr_setidealprocessor_np` and `pthread_setidealprocessor_np` (plus `get` variants).

Here is a listing of the known (and slightly less known - nonstandard) pthread functions we could reuse:
| Function | Usage | Link |
|-|-|-|
| pthread_attr_setschedparam | Set the priority attribute value | https://man7.org/linux/man-pages/man3/pthread_attr_getschedparam.3p.html |
| pthread_attr_getschedparam | Get the priority attribute value | https://man7.org/linux/man-pages/man3/pthread_attr_getschedparam.3p.html |
| pthread_setschedparam | Set the thread's priority value | https://man7.org/linux/man-pages/man3/pthread_setschedparam.3.html |
| pthread_getschedparam | Get the thread's priority value | https://man7.org/linux/man-pages/man3/pthread_setschedparam.3.html |
| pthread_attr_setaffinity_np | Set the affinity mask attribute value | https://man7.org/linux/man-pages/man3/pthread_attr_setaffinity_np.3.html |
| pthread_attr_getaffinity_np | Get the affinity mask attribute value | https://man7.org/linux/man-pages/man3/pthread_attr_setaffinity_np.3.html |
| pthread_setaffinity_np | Set the thread's affinity mask value | https://man7.org/linux/man-pages/man3/pthread_setaffinity_np.3.html |
| pthread_getaffinity_np | Get the thread's affinity mask value | https://man7.org/linux/man-pages/man3/pthread_setaffinity_np.3.html |

Note that the `pthread_attr_setaffinity_np` function, if we used it, would only apply after creating the thread since we don't have an API to set it during creation. See also https://stackoverflow.com/a/35278467.

I think we should keep the thread builder extension trait added in this PR, but we could also consider adding APIs for modifying the priority, ideal processor, and affinity mask of existing threads (might want to keep these in `std::os::horizon::thread`).

**Comment by Meziu** [comment]
Great. This PR will be merged _after_ the `std` PR, so you can make all the changes needed. Just rename them in libc and we are good to go.

**Comment by AzureMarker** [comment]
We could also merge this PR when it's ready to get it out of the way, and make a separate branch for the std PR. We may need to clean up the commit history or split it into two PRs, so making a separate branch might be necessary either way.

**Comment by Meziu** [comment]
Yeah we should squash some stuff and clean up the repo. Let’s make a branch.

File: library/std/src/sys/unix/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/unix/thread.rs]: This has been updated to use a standard API for priority, but the affinity/"ideal processor" API isn't standard anywhere else, so I've made an explicitly non-portable (np) version.

**Comment by AzureMarker** [comment]
By the way, this thread is interesting (even though it never got properly answered):
https://gbatemp.net/threads/can-horizon-indepently-relocate-threads-to-cpus-in-their-affinity-masks.556096/

I don't have a New 3DS/2DS to test with (to see if it balances across multiple app cores), but it does seem a little odd that there's the whole affinity mask feature if it's not really used by the kernel after initial scheduling.

**Comment by AzureMarker** [comment]
I did a quick test to try migrating a thread from sys core to app core (and vice versa) via `svcSetThreadIdealProcessor`. It fails with an error (result), level = RL_FATAL, summary = RS_NOTSUPPORTED, description = RD_NOT_IMPLEMENTED. This happens even if the thread's affinity mask allows it to exist on either app or sys core (mask = 0x3).

Due to this, I don't think it makes sense to add this feature (setting the ideal processor after spawning) to std. I'm not sure if it's worthwhile to add the get version either, since that could return stuff like -2 or -1 which don't tell you what processor it is actually on. The only API I know of to do that is `svcGetProcessorID`, and that only gets the current thread's value.

So basically, (I think) we should just have functions to get the current thread's priority and processor ID in std, but not the ideal thread or affinity mask.

File: library/std/src/os/horizon/thread.rs
Comment (AzureMarker) [inline library/std/src/os/horizon/thread.rs]: Naming wise, should this be just `processor` or still `ideal_processor`? Using `affinity` might be confusing since there's a separate affinity mask, and `ideal_processor` seems a little long/wordy.

This naming should probably be the same across std and the nonstandard libc functions.

Comment (ian-h-chamberlain) [reply library/std/src/os/horizon/thread.rs]: I think `ideal_processor` is fine, since there is some reference to this wording in `libctru` already, but it could also be `processor_id` to match [the libctru argument](https://libctru.devkitpro.org/svc_8h.html#a3f62a8f1822c6a78fab2e2cb03c46893).

**Comment by ian-h-chamberlain** [comment]
> I don't have a New 3DS/2DS to test with (to see if it balances across multiple app cores), but it does seem a little odd that there's the whole affinity mask feature if it's not really used by the kernel after initial scheduling.

I have a N3DS if you have an example you'd like me to test – if those functions are implemented on the new processors perhaps it would be worth keeping the implementation?

**Comment by AzureMarker** [comment]
@ian-h-chamberlain nice, I was testing with a variant of this example:
https://github.com/Meziu/ctru-rs/pull/46/commits/353b6c1c77489a0c56e83418a67d572964b30bc4

I think you have to enable the new cores through some mechanism. This is what the docs say:
> Processor #2 is New3DS exclusive. Normal applications can create threads on this core if the exheader kernel flags bitmask has 0x2000 set.

Then just change the ~~ideal processor~~ processor ID setting of the thread to -1 and try to move the thread to another core with something like this:
```rust
println!("Moving to app core 2");
let result =
    unsafe { ctru_sys::svcSetThreadIdealProcessor(ctru_sys::CUR_THREAD_HANDLE, 2) };
if ctru_sys::R_FAILED(result) {
    println!("result level = {}", ctru_sys::R_LEVEL(result));
    println!("result summary = {}", ctru_sys::R_SUMMARY(result));
    println!("result description = {}", ctru_sys::R_DESCRIPTION(result));
    return;
}

println!("On app core 2");
print_affinity_mask("app 2 thread");
print_thread_id("app 2 thread");
print_thread_list();
```

File: library/std/src/os/horizon/thread.rs
Comment (AzureMarker) [reply library/std/src/os/horizon/thread.rs]: I think I'll switch to `processor_id` to make things a little less... exotic.

Comment (AzureMarker) [reply library/std/src/os/horizon/thread.rs]: Opened https://github.com/Meziu/libc/pull/10 and https://github.com/Meziu/pthread-3ds/pull/11

**Comment by Meziu** [comment]
Even if it was technically possible to move threads around, it looks like the 3DS is the worst system to do so. There are 2 *VERY* different cores on a normal 3DS, and 4 (yet one is totally unusable) on a New3DS. I have never even thought of trying: why would you try to move a thread that’s supposed to be preemptive to a core that isn’t? Or try to move a thread to a core some systems don’t even have? It usually ends up being: common threads on core#0, and something important on core#1. I don’t think it matters.

File: library/std/src/os/horizon/thread.rs
Comment (ian-h-chamberlain) [inline library/std/src/os/horizon/thread.rs]: Didn't think of this before, but would `BuilderExt` be more consistent with `thread::Builder`? Not big deal either way I think

**Comment by ian-h-chamberlain** [comment]
> Even if it was technically possible to move threads around, it looks like the 3DS is the worst system to do so. There are 2 _VERY_ different cores on a normal 3DS, and 4 (yet one is totally unusable) on a New3DS. I have never even thought of trying: why would you try to move a thread that’s supposed to be preemptive to a core that isn’t? Or try to move a thread to a core some systems don’t even have? It usually ends up being: common threads on core#0, and something important on core#1. I don’t think it matters.

Ok, I got a chance to test this on my device and it seems that it's not supported anyway.
I got an error code with summary code "Not supported", description "Not implemented", so I think we can safely ignore this functionality for the reasons you describe and the fact that it doesn't work anyway.

File: library/std/src/os/horizon/thread.rs
Comment (AzureMarker) [reply library/std/src/os/horizon/thread.rs]: I thought of that when I made the trait, but thinking about it again, I think it would work. My concern was that it would look odd as `std::os::horizon::BuilderExt`, but in fact it's `std::os::horizon::thread::BuilderExt` which looks better. I'll rename it.

File: library/std/src/sys/sgx/thread.rs
Comment (AzureMarker) [reply library/std/src/sys/sgx/thread.rs:110]: We got some good feedback on the Zulip thread. The action items are:
1. Make a PR (or multiple) with the changes we already have now.
2. Make the thread changes in a separate PR.

See #6 for followups on item 1.

**Comment by AzureMarker** [comment]
I think the PR is ready to merge, whenever we want to do that. For the std PRs we can cherry-pick changes to separate branches, so I don't think that would block merging this PR.

**Review by Meziu — APPROVED** [review]

**Comment by AzureMarker** [comment]
@Meziu Are you waiting to merge this? What's the plan?

**Comment by Meziu** [comment]
> @Meziu Are you waiting to merge this? What's the plan?

Aren’t there some memory issues related with #15 and https://github.com/Meziu/ctru-rs/issues/48 ? I thought the code wasn’t final yet.

File: library/std/src/os/horizon/thread.rs
Comment (Meziu) [inline library/std/src/os/horizon/thread.rs]: Could you remove the link to the syscall? 

Maybe we should make this available to the user too via `pthread`?

File: library/std/src/os/horizon/thread.rs
Comment (Meziu) [inline library/std/src/os/horizon/thread.rs]: Remove the reference to the ExHeader kernel flag. It isn't in the `std` scope to change that (that only really matters when building CIAs, which is not something the `std` should worry about).

Something like:
```suggestion
    /// * Processor #2 is New3DS exclusive. Normal applications can create
    ///   threads on this core only if the built application has proper external setup.
```

**Comment by AzureMarker** [comment]
> > @Meziu Are you waiting to merge this? What's the plan?
> 
> Aren’t there some memory issues related with #15 and [Meziu/ctru-rs#48](https://github.com/Meziu/ctru-rs/issues/48) ? I thought the code wasn’t final yet.

Oh yeah, not sure how I forgot that blocker. I guess I thought of this code as "done", and any future changes would need another PR. With that in mind though, I do have one small change planned for this PR before it merges (along with the review comments).

File: library/std/src/os/horizon/thread.rs
Comment (AzureMarker) [reply library/std/src/os/horizon/thread.rs]: We could add that functionality in a separate PR. I don't think we should expose too much via std, just enough to unblock people who want to use libraries that rely on std threads. The other 3ds specific stuff can be done via ctru-rs for now.

**Comment by AzureMarker** [comment]
Once https://github.com/Meziu/pthread-3ds/pull/14 merges, I think this could merge too. The small change I wanted to make was https://github.com/Meziu/rust-horizon/pull/10/commits/3b5e287a9ec06158191cea81645870ef8a1e8677.

**Comment by AzureMarker** [comment]
Thanks @Meziu and @ian-h-chamberlain for reviewing and keeping up with this PR!

**Pull request merged**
